electroncash/lns.py

```diff
@@ -0,0 +1,607 @@
+                        self.v_by_name[item.name].add(item)
+                        self.v_by_addr[item.address].add(item)
+                    self.save(True)
+                    l = self._names_in_flight.pop(key, [])
+                ct = 0
+                for success_cb, error_cb in l:
+                    if success_cb:
+                        success_cb(pb)
+                        ct += 1
+                if debug: self.print_error(f"verify_name_asynch: called {ct} success callbacks for #{key}")
+            else:
+                on_error(exc[-1])
+        with self.lock:
+            l = self._names_in_flight[key]
+            l.append((success_cb, error_cb))
+            if len(l) == 1:
+                if debug: self.print_error(f"verify_name_asynch: initiating new lookup_asynch_all on #{key}")
+                lookup_asynch_all(name=name, success_cb=on_success, error_cb=on_error, timeout=timeout, debug=debug)
+            else:
+                if debug: self.print_error(f"verify_name_asynch: #{key} already in-flight, will just enqueue callbacks")
+
+    ############################
+    # UI / Prefs / Convenience #
+    ############################
+
+    def get_address_default(self, infos : List[Info]) -> Info:
+        ''' Returns the preferred Info object for a particular address from
+        a given list. `infos' is a list of Info objects pertaining to a
+        particular address (they should all pertain to said address, but this
+        is not checked). '''
+        if infos:
+            last = infos[-1]
+            d = self.wallet.storage.get('lns_address_defaults')
+            if isinstance(d, dict) and isinstance(last.address, Address):  # sanity check, .address may not always be Address but may be UnknownAddress
+                tup = d.get(last.address.to_storage_string())
+                if isinstance(tup, (tuple, list)) and len(tup) == 3:
+                    name, number, chash = tup
+                    if isinstance(name, str) and isinstance(number, (int, float)) and isinstance(chash, str):
+                        # find the matching one in the list
+                        for info in infos:
+                            if (name.lower(), number, chash) == (info.name.lower(), info.number, info.collision_hash):
+                                return info
+            # just return the latest one if no default specified
+            return last
+
+    def set_address_default(self, info : Info):
+        ''' Set the default LNS Name for a particular address. Pass the Info
+        object pertaining to the LNS Name / Address in question. '''
+        if not isinstance(info.address, Address):
+            self.print_error("Warning: Info object does not have an Address", info)
+            return
+        d = self.wallet.storage.get('lns_address_defaults', {})
+        addr_str = info.address.to_storage_string()
+        new_value = [info.name, info.number, info.collision_hash]
```

Comment:
this looks like cashaccount info? imho we should not be saving anything other than the lns names for contacts, as the addresses they are pointing to may change even while wallet is open (long running wallets).

Reply:
This is the most critical item mentioned in this PR feedback. The whole point of LNS is to allow people to dynamically change their address through the contract. The address itself should not be saved.

Reply:
Agree, this can be removed.
Regarding address change, there is a 10 minute timer which forces the LNS cache revalidation. Also whenever you put an LNS name in a `send` tab it fetches from cache, it can be forced to always fetch from network, though. Also on each app start if addresses change the contacts get replaced with new address.

Reply:
removed the relevant code. not resolving this convo yet

Reply:
tested, no faults